### PR TITLE
Upgrade to GASNet-EX 2018.9.0 

### PR DESCRIPTION
**Pull request by ronawho**
GASNet-EX is an evolution of GASNet-1 for exascale. It has potential for some
dramatic performance benefits as well as new feature like support for network
atomics. However, those require using new APIs and here we're still just
calling the GASNet-1 compatible API so there shouldn't be any major performance
differences.

This release of EX does not have support for the psm or ofi conduits. To help
with this I've added an explicit check for these configs and we'll bark if
somebody tries to use them with a message asking them to contact us. If these
configurations are really important to somebody we can send them a GASNet-1
tarball and they should just be able to drop that in. To support that, I left
the gasnetrun_{psm,ofi} launchers and the chplenv logic for them alone. I did
remove all user-facing documentation for them though.

 - Correctness testing:
   - [X] full udp-everything testing
   - [X] full udp-fast testing
   - [X] full aries testing
   - [X] hellos for gemini
   - [x] hellos for remaining configs/launchers we can test
        - ibv, mpi, smp

 - Performance testing:
   - [x] full aries performance testing

 - Other:
   - [x] figure out deprecation story for psm/ofi
   - [x] documentation updates?
   - [x] verify I applied upstream patches correctly
        - tested mpi with Open MPI 3.1 and verified placement
        - tested aries on heterogenous node types
   - [x] add missing nightly testing: https://github.com/chapel-lang/chapel/issues/11537
   - [x] looks like the license has changed with EX, verify it is still BSD-like
        - new license still says "BSD-style license"

Closes https://github.com/chapel-lang/chapel/issues/11696

### CODE REVIEW

**Comment by ronawho** [comment]
@dmk42 mostly an FYI that I dropped support for the aarch64 cross compile script. 

@gbtitus this is easiest to review commit-by-commit (skipping the actual EX upgrade)

**Comment by ronawho** [comment]
Note that psm was added by a user in https://github.com/chapel-lang/chapel/pull/5843, so I tried pinging them to see if that config was still important, but did not hear back -- https://github.com/chapel-lang/chapel/pull/5843#issuecomment-440477744

File: third-party/gasnet/README [hunk 1]
```diff
@@ -18,9 +18,9 @@ The modifications that we have made to the official GASNet release are
 as follows:
 
 * Pulled in an upstream fix for mpi-spawner -N support for Open MPI 3.x
-   - https://bitbucket.org/berkeleylab/gasnet/commits/e9088bc72
+   - https://bitbucket.org/berkeleylab/gasnet-ex/commits/bdd352a5c
```
Comment (gbtitus) [inline third-party/gasnet/README]: These commit IDs don't match the ones given in the commit messages for the cherry-picks themselves (not our commit IDs, I wouldn't expect that, but they also don't match the commit IDs we say we're cherry-picking from).  Should they?

Comment (ronawho) [reply third-party/gasnet/README]: For our https://github.com/chapel-lang/chapel/pull/11874/commits/52cad0e218f30eb3f7e38d4ccfc00537ddfb8a8f (which cherry-picked https://bitbucket.org/berkeleylab/gasnet-ex/commits/bdd352a5c) you're seeing something like: 

```
(cherry picked from commit e9088bc72825652a6221a778cbc5617911520f66)
```

That's from the EX commit message itself. They cherry picked from the GASNet-1 or dev branch onto EX branch, and then I cherry-picked it from the EX branch. (Note that that matches the value we used to cherry-pick too.)

**Comment by bonachea** [comment]
Note that an ofi-conduit for GEX is in our FY19 roadmap, which is another good reason to leave the Chapel-side ofi code in place for that eventual upgrade.

**Comment by ronawho** [comment]
Workaround for psm/ofi users of the Chapel 1.19 release:

```sh
# Ignore our errors about ofi/psm not being supported
export CHPL_GASNET_ALLOW_BAD_SUBSTRATE=true

# Replace GASNet-EX with GASNet-1
wget https://gasnet.lbl.gov/download/GASNet-1.32.0.tar.gz
tar xzf GASNet-1.32.0.tar.gz
rm -rf $CHPL_HOME/third-party/gasnet/gasnet-src
mv GASNet-1.32.0 $CHPL_HOME/third-party/gasnet/gasnet-src
```

**Pull request merged**